File: backend/dao/controls.py
"""DAO for controls dataset."""
from typing import Dict, Any, Optional
from dao.base import BaseDAO


class ControlsDAO(BaseDAO):
    """Data Access Object for controls dataset."""

    def __init__(self):
        super().__init__(dataset_name="controls")

    def get_details(self, control_id: str) -> Optional[Dict[str, Any]]:
        return super().get_details(control_id)

    async def trigger_controls_taxonomy(
        self,
        control_id: str,
        session_id: str,
        user_id: str,
        refresh: bool = False,
    ) -> Dict[str, Any]:
        return await self.trigger_ai_function(
            control_id,
            "controls_taxonomy",
            session_id,
            user_id,
            refresh,
        )

    async def trigger_root_cause(
        self,
        control_id: str,
        session_id: str,
        user_id: str,
        refresh: bool = False,
    ) -> Dict[str, Any]:
        return await self.trigger_ai_function(
            control_id,
            "root_cause",
            session_id,
            user_id,
            refresh,
        )

    async def trigger_enrichment(
        self,
        control_id: str,
        session_id: str,
        user_id: str,
        refresh: bool = False,
    ) -> Dict[str, Any]:
        return await self.trigger_ai_function(
            control_id,
            "enrichment",
            session_id,
            user_id,
            refresh,
        )

File: frontend/src/api/controls.js
// API service for controls dataset
import axiosClient from './axiosClient';

const triggerControlsTaxonomy = async (controlId, sessionId, userId, refresh = false) => {
  const response = await axiosClient.post(
    `/api/controls/${controlId}/controls-taxonomy`,
    { session_id: sessionId, user_id: userId },
    { params: { refresh } }
  );
  return response.data;
};

const triggerRootCause = async (controlId, sessionId, userId, refresh = false) => {
  const response = await axiosClient.post(
    `/api/controls/${controlId}/root-cause`,
    { session_id: sessionId, user_id: userId },
    { params: { refresh } }
  );
  return response.data;
};

const triggerEnrichment = async (controlId, sessionId, userId, refresh = false) => {
  const response = await axiosClient.post(
    `/api/controls/${controlId}/enrichment`,
    { session_id: sessionId, user_id: userId },
    { params: { refresh } }
  );
  return response.data;
};

const controlsAPI = {
  search: async (id, limit = 100) => {
    const response = await axiosClient.get('/api/controls', {
      params: { id, limit }
    });
    return response.data;
  },

  getDetails: async (controlId) => {
    const response = await axiosClient.get(`/api/controls/${controlId}/details`);
    return response.data;
  },

  list: async (offset = 0, limit = 100) => {
    const response = await axiosClient.get('/api/controls/list', {
      params: { offset, limit }
    });
    return response.data;
  },

  aiTriggers: {
    controls_taxonomy: triggerControlsTaxonomy,
    root_cause: triggerRootCause,
    enrichment: triggerEnrichment
  },

  meta: {
    primaryFunction: 'controls_taxonomy',
    titleField: 'control_title',
    typeField: 'key_control',
    themeField: 'risk_theme',
    subthemeField: 'risk_subtheme',
    functionLabels: {
      controls_taxonomy: 'Controls Taxonomy',
      root_cause: 'Root Cause',
      enrichment: 'Enrichment'
    }
  }
};

export default controlsAPI;
